enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,182 @@
+
+### Kubelet
+
+Openshift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are passed externally from the `Infrastructure` resource on https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357, where the support for `external` cloud provider should be added. For each platform with `external` configuration, the `cloud-config` flag should not be set.
+
+### Kube-api-server
+
+No changes for the current configuration are needed
+
+### Kube-controller-manager
+
+TBD
+
+#### Upgrade/Downgrade strategy
+
+Upstream [proposal](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#motivation) describes main pinpoints for migration between in-tree and out-of-tree for HA clusters. They propose implementation for cross-namespace leadership delegation, which would help Openshift to preserve HA during upgrages.
+
+In case the cluster does not require to be HA, then upgrading under CVO management, which will lead to two simultanious replicas of `cloud-controllers` running at the same time at some moment is not a problem. Each provider, therefore supports multiple options to run the `cloud-controller`:
+
+1) As a `DaemonSet`. This is the HA option, where the leader-election on the running pod ensuring a single replica is serving cluster at any point in time.
+2) As a `Pod`. For Openshift this option is not sufficient, so non-HA environment will run this pod under `Deployment` or as a static resource managed from the `Node` filesystem in the `openshift-cloud-controller` namespace.
+
+#### Examples
+
+##### AWS
+
+Main repository: https://github.com/kubernetes/cloud-provider-aws/
+Sample manifests: https://github.com/kubernetes/cloud-provider-aws/tree/master/manifests
+
+To try external cloud-controller-manager, siply run:
+```bash
+oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/rbac.yaml  
+oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/aws-cloud-controller-manager-daemonset.yaml
+```
+
+##### Azure
+
+Main repository: https://github.com/kubernetes-sigs/cloud-provider-azure
+Sample manifests: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/examples/out-of-tree
+
+##### GCP
+
+Main repository: https://github.com/kubernetes/cloud-provider-gcp
+Sample manifests: https://github.com/kubernetes/cloud-provider-gcp/tree/master/deploy
+
+##### vSphere
+
+Main repository: https://github.com/kubernetes/cloud-provider-vsphere
+Sample manifests: https://github.com/kubernetes/cloud-provider-vsphere/tree/master/manifests/controller-manager
+
+### Test Plan
+
+- Each provider at the time of migration should have a working CI implemetation, which will assist in testing the provisioning with out-of-tree support enabled.
```

Comment:
Do you think existing CI should be sufficient to cover this, or do we need to add new things? (The answer to this should be reflected into the doc)

Reply:
I would assume that having CI tests to execute upgrades on each provider onto cluster configuration managed externally, will be enough to confirm this is working. We could gain this for free on the final PR implementing the default `external` option in the installer, but probably the better way (or an additional milestone) will be to replicate these steps in CI for the PRs published in the proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository.

Reply:
Proposed annotation on the `Infrastructure` resource might help with CI here too.